implementing optional content radiobutton groups

agrahn opened the pull request:
This PR implements OC radio button groups. It could resolve #18823.

The suggested code parses the `/RBGroups` entry in the optional content configuration dict and adds the property `myRbGroups` to instances of the `OptionalContentGroup` class. `myRbGroups` takes an array of `Set`s , where each `Set` instance represents a radio button group the  `OptionalContentGroup` instance is a member of. Such a `Set` instance contains all OCG ids within the corresponding RB group. The RB groups an OCG is associated with are processed when its visibility is set to `true`, as required by the PDF spec.

agrahn commented:
@Snuffleupagus I updated the PR with your version of `pdf_layer_viewer.js` from `master`. Only two changed files remain for review :).

Snuffleupagus commented on src/core/catalog.js:
```diff
@@ -486,17 +486,17 @@ class Catalog {
       }
       const groups = [];
-      const groupRefs = new RefSet();
+      const groupsMap = new Map();
```
*This comment applies in multiple other spots below as well.*

Please note that it's not generally safe/correct to store `Ref`-instances in `Set`s and `Map`s, which is why we have `RefSet` respectively `RefSetCache` for that purpose; see [classes here](https://github.com/mozilla/pdf.js/blob/35a9a6a7b34030397c2215ebd3d17064ac53887e/src/core/primitives.js#L318-L391).

In this case you thus want the following instead.
```suggestion
      const groupRefCache = new RefSetCache();
```

agrahn replied:
See the comment below.

Snuffleupagus commented on src/core/catalog.js:
```diff
@@ -517,6 +517,7 @@ class Catalog {
         print: null,
         view: null,
       },
+      myRbGroups: [],
```
Is there a reason for prefacing this name with "my", or can the following be used instead?
```suggestion
      rbGroups: [],
```

agrahn replied:
I chose that name because the array property does not contain all RB groups in the document, but only those that the OCG belongs to.

Snuffleupagus replied:
But the entry is only for the *current* group, so the "my" part really is redundant.

Snuffleupagus commented on src/core/catalog.js:
```diff
@@ -638,10 +639,39 @@ class Catalog {
       return { name: stringToPDFString(nestedName), order: nestedOrder };
     }
 
+    function parseRBGroups(rbgrps) {
```
*Similar comments also applies elsewhere in the patch.*

Shortening this doesn't seem entirely necessary (nor all that readable).
```suggestion
    function parseRBGroups(rbGroups) {
```

Snuffleupagus commented on src/core/catalog.js:
```diff
@@ -638,10 +639,39 @@ class Catalog {
+        return null;
+      }
+
+      // iterate over RB groups (literal arrays or refs thereof)
```
*This comment applies in multiple other spots throughout the patch as well.*

To improve readability please remember that comments should be complete sentences, i.e. beginning with an upper-case letter and ending with a period.

```suggestion
      // Iterate over RB groups (literal arrays or refs thereof).
```

Snuffleupagus commented on src/core/catalog.js:
```diff
@@ -638,10 +639,39 @@ class Catalog {
+
+      // iterate over RB groups (literal arrays or refs thereof)
+      for (const value of rbgrps) {
+        const rbGroup = value instanceof Ref ? xref.fetchIfRef(value) : value;
```
This is actually redundant, see [here](https://github.com/mozilla/pdf.js/blob/35a9a6a7b34030397c2215ebd3d17064ac53887e/src/core/xref.js#L794-L799), and either of the following options are correct:
```suggestion
        const rbGroup = value instanceof Ref ? xref.fetch(value) : value;
```
or

```suggestion
        const rbGroup = xref.fetchIfRef(value);
```

agrahn replied:
I will use the second suggestion.

Snuffleupagus commented on src/core/catalog.js:
```diff
@@ -638,10 +639,39 @@ class Catalog {
+      // iterate over RB groups (literal arrays or refs thereof)
+      for (const value of rbgrps) {
+        const rbGroup = value instanceof Ref ? xref.fetchIfRef(value) : value;
+        // ignore wrong (non-array) refs and empty arrays
```
This comment seems a bit redundant here, since it's quite obvious what the next lines does.

agrahn replied:
Ok, will delete this comment.

Snuffleupagus commented on src/display/optional_content_config.js:
```diff
@@ -229,12 +230,26 @@ class OptionalContentConfig {
+    // if my visibility is about to be set to `true' and if I belong to one or
+    // more radiobutton groups, hide all other OCGs in these radiobutton groups,
+    // provided that radio-button state relationships are to be preserved
+    if (visible && group.myRbGroups.length && preserveRB) {
```
Let's re-order this slightly, such that we do the "simpler" checks first.
```suggestion
    if (preserveRB && visible && group.myRbGroups.length) {
```

Snuffleupagus reviewed (changes requested):
Thanks for the patch; based on an initial look I've left a number of inline comments.

Please remember to include the information from https://github.com/mozilla/pdf.js/pull/18825#issue-2557588601 in the *commit message* as well, such that the patch is easier to understand also on the Git command line.

agrahn commented:
Thank you, @Snuffleupagus, for your comments. I am going to work through all of them. Is it OK to do the code changes locally and to force-push them afterwards?

agrahn commented on src/core/catalog.js:
```diff
@@ -486,17 +486,17 @@ class Catalog {
+        groupsMap.set(groupRef, group);
       }
-      config = this.#readOptionalContentConfig(defaultConfig, groupRefs);
+      config = this.#readOptionalContentConfig(defaultConfig, groupsMap);
```
The reason why I am using `Map` instead of `RefSetCache` is that I need its `keys()` method which the latter does not provide:  `groupsMap` is passed to `#readOptionalContentConfig` on line 499. Then, in the local function `parseOrder`, I want to iterate over the keys when the `hiddenGroups` array is filled.

Snuffleupagus replied:
https://github.com/mozilla/pdf.js/blob/e25832c43fe4d945b0773aef9abf7d8a6325c34f/src/core/primitives.js#L386-L390

agrahn replied:
I need some more help on this as I still don't know how to translate the `Map`-based to a `SetRefCache-based` formulation (function parseOrder) . None of 

````
const groupRef of contentGroupsMap.keys() -->  const groupRef of groupRefCache.items().keys()
const groupRef of contentGroupsMap.keys() -->  const groupRef of groupRefCache.items()
````
works.

Snuffleupagus replied:
```javascript
for (const [groupRef] of contentGroupRefCache.items()) {
  ...
}
```

agrahn replied:
Thanks! I did it slightly differently in the meantime. I updated the code once more using your solution.

agrahn commented:
> Please remember to include the information from [#18825 (comment)](https://github.com/mozilla/pdf.js/pull/18825#issue-2557588601) in the _commit message_ as well, such that the patch is easier to understand also on the Git command line.

That's a lot of text. I thought a commit message should fit on a single line. But I will do so when appropriate.

Snuffleupagus reviewed (changes requested):
This also needs tests, and I'd suggest adding a couple of reference tests for this.
 - Add the PDF document that you attached in https://github.com/mozilla/pdf.js/issues/18823#issue-2557152518 to the `test/pdfs/` folder, but name it `issue18823.pdf`.
 - Run `node test/add_test.mjs test/pdfs/issue18823.pdf` in the root of your local PDF.js Git-clone.
 - Open the `test/test_manifest.json` file and find the newly added test-case. Here we also want to add *another* test-case for a non-default optionalContent-state.

To make this process slightly less daunting I've done these steps in the following diff, which you're obviously welcome to copy and include as-is in your patch: https://github.com/mozilla/pdf.js/compare/master...Snuffleupagus:pdf.js:issue-18823-test

agrahn commented:
> This also needs tests, and I'd suggest adding a couple of reference tests for this.
> 
>    * Add the PDF document that you attached in [[Feature]: Optional Content Radio Button Groups #18823 (comment)](https://github.com/mozilla/pdf.js/issues/18823#issue-2557152518) to the `test/pdfs/` folder, but name it `issue18823.pdf`.
> 
>    * Run `node test/add_test.mjs test/pdfs/issue18823.pdf` in the root of your local PDF.js Git-clone.
> 
>    * Open the `test/test_manifest.json` file and find the newly added test-case. Here we also want to add _another_ test-case for a non-default optionalContent-state.
> 
> 
> To make this process slightly less daunting I've done these steps in the following diff, which you're obviously welcome to copy and include as-is in your patch: [master...Snuffleupagus:pdf.js:issue-18823-test](https://github.com/mozilla/pdf.js/compare/master...Snuffleupagus:pdf.js:issue-18823-test)

Thanks a lot for this, I merged the test case into the patch. I just don't quite understand what is to be done regarding the other test case for a non-default OC state (third item in your comment).

Snuffleupagus commented:
> Thanks a lot for this, I merged the test case into the patch. I just don't quite understand what is to be done regarding the other test case for a non-default OC state (third item in your comment).

That patch already includes all of those steps, however I just wanted to highlight that the script can only create "simple" entries in the `test/test_manifest.json` file.

 - The script I mentioned will generate this part automatically: https://github.com/mozilla/pdf.js/commit/3bacbf65cd94358d2bd2b1a2de815775269913ff#diff-9bc50ce4ba8a4ba8d322114a3cc05fa62005a40b52223059200ca07087dd5a61R6878-R6884

 - The third point cannot be automatically generated, however I've manually added it in that diff here: https://github.com/mozilla/pdf.js/commit/3bacbf65cd94358d2bd2b1a2de815775269913ff#diff-9bc50ce4ba8a4ba8d322114a3cc05fa62005a40b52223059200ca07087dd5a61R6885-R6894

---

It seems that the PR currently contains unrelated commits, maybe a rebase gone wrong?

agrahn committed:
Optional Content (OC) radiobutton (RB) groups implemented. Resolves #18823.

The code parses the /RBGroups entry in the OC configuration dict and adds the property `rbGroups' to instances of the OptionalContentGroup class. rbGroups takes an array of Sets, where each Set instance represents an RB group the OptionalContentGroup instance is a member of. Such a Set instance contains all OCG ids within the corresponding RB group. RB groups an OCG is associated with are processed when its visibility is set to true, as required by the PDF spec.

agrahn commented:
>    * The third point cannot be automatically generated, however I've manually added it in that diff here: [3bacbf6#diff-9bc50ce4ba8a4ba8d322114a3cc05fa62005a40b52223059200ca07087dd5a61R6885-R6894](https://github.com/mozilla/pdf.js/commit/3bacbf65cd94358d2bd2b1a2de815775269913ff#diff-9bc50ce4ba8a4ba8d322114a3

OK, so nothing else needs to be done regarding this.

> It seems that the PR currently contains unrelated commits, maybe a rebase gone wrong?

I hope the commit history is in sync again. Going to work on the other changes now.

agrahn commented on src/core/catalog.js:
```diff
@@ -642,6 +643,32 @@ class Catalog {
       parsedOrderRefs = new RefSet(),
       MAX_NESTED_LEVELS = 10;
 
+    // Parse RBGroups entry.
+    (rbGroups => {
```
Because the function is local + only called once while loading a PDF and its return value is not used, I thought an IIFE would be in order. However, I can change it back to a named function that is called subsequently.

agrahn commented on src/core/catalog.js:
```diff
@@ -642,6 +643,32 @@ class Catalog {
+      }
+
+      return null;
+    })(config.get("RBGroups"));
```
IIFE converted to named function.

agrahn commented on src/display/optional_content_config.js:
```diff
@@ -229,12 +230,26 @@ class OptionalContentConfig {
+      for (const rbGrp of group.rbGroups) {
+        for (const otherId of rbGrp) {
+          if (otherId !== id) {
+            this.#groups.get(otherId)._setVisible(INTERNAL, false, true);
```
OK

agrahn commented on src/display/optional_content_config.js:
```diff
@@ -229,12 +230,26 @@ class OptionalContentConfig {
+    // If my visibility is about to be set to `true' and if I belong to one or
+    // more radiobutton groups, hide all other OCGs in these radiobutton groups,
+    // provided that radiobutton state relationships are to be preserved.
+    if (visible && preserveRB && group.rbGroups.length) {
```
OK

agrahn commented on src/display/optional_content_config.js:
```diff
@@ -229,12 +230,26 @@ class OptionalContentConfig {
     }
+
+    // If my visibility is about to be set to `true' and if I belong to one or
+    // more radiobutton groups, hide all other OCGs in these radiobutton groups,
```
OK

agrahn commented on src/core/catalog.js:
```diff
@@ -642,6 +643,32 @@ class Catalog {
+        if (!Array.isArray(rbGroup) || !rbGroup.length) {
+          continue;
+        }
+
+        const parsedRbGroup = new Set();
```
The plural would be misleading here; `parsedRbGroup` indeed represents a *single* radio button group. This `Set` variable collects the OCG refs that belong to the radio button group that is currently parsed.

agrahn commented on src/core/catalog.js:
```diff
@@ -642,6 +643,32 @@ class Catalog {
+    // Parse RBGroups entry.
+    (rbGroups => {
+      if (!Array.isArray(rbGroups)) {
+        return null;
```
OK

Snuffleupagus commented:
/botio-linux preview

Snuffleupagus commented on src/core/catalog.js:
```diff
@@ -486,17 +486,17 @@ class Catalog {
       }
-      config = this.#readOptionalContentConfig(defaultConfig, groupRefs);
+      config = this.#readOptionalContentConfig(defaultConfig, groupRefCache);
       config.groups = groups;
```
Do we actually need the separate `groups` variable now, or could we just do the following here?
```suggestion
      config.groups = [...groupRefCache];
```

Snuffleupagus commented:
/botio test

Snuffleupagus commented on src/core/catalog.js:
```diff
@@ -638,10 +639,39 @@ class Catalog {
+          if (
+            ref instanceof Ref &&
+            groupRefCache.has(ref) &&
+            !parsedRbGroup.has(ref)
```
Unfortunately I don't believe that this last check works, since `parsedRbGroup` contains strings while `ref instanceof Ref` here.

agrahn replied:
I cannot imagine that the rather random pixel discrepancies of the failed testcases have anything to do with the added OC-related code of the patch. Only the PDF of issue2442 contains an `OCProperties` dictionary. While it has an `RBGroups` sub-key, the value thereof is an empty array. So, even here, the code listed above is never reached and thus cannot have an impact on the rendering. Given that the discrepancies of colour components are not larger than `1`, they might be caused by round-off error. Are, perhaps, the reference images outdated?

Snuffleupagus replied:
There's unfortunately a number of intermittent issues in the reference tests, so that's usually not anything to worry about and thus *not* what I commented on above!

---

The point here is that there's a logical "error" in the patch, since `!parsedRbGroup.has(ref) === true` will always hold and that check is thus pointless as-is. Did you actually mean the following?
```suggestion
            !parsedRbGroup.has(ref.toString())
```

agrahn replied:
Oh, apologies, I completely misread your former comment! You are completely right, the test as of now does not make sense at all. Of course, `ref.toString()` is needed here! Maybe, it would have been safer to use `RefSet` instead of `Set` for `parsedRbGroup`, where conversion to string is internally handled. But I don't now how to iterate over `RefSet` values in a `for...of` loop. This is needed in `src/display/optional_content_config.js`, line 245. 

How can I make use of the `Symbol.iterator` defined in class `RefSet`?

Snuffleupagus reviewed (approved):
r=me, thank you.

Snuffleupagus commented:
/botio makeref
